### ISSUE TITLE
Auth error with libmamba solver

### ISSUE THREAD

**Issue by VitaliyaIoffe** [conda/conda-libmamba-solver#216]

**Comment by travishathaway** [comment]
Hi @VitaliyaIoffe,

Thanks for filing this issue. Just to let you know, we have moved it over to the https://github.com/conda/conda-libmamba-solver repository. If you run into any more issues using the new solver, please them here.

The team will take a look at this at their nearest convenience.

**Comment by jaimergp** [comment]
Hey @VitaliyaIoffe, I am looking now into this.

We had similar reports for previous versions (see #146 and #108). Can you confirm you are seeing this error with the latest version (`conda-libmamba-version 23.5.0`)? If that's the case, do you see this if you export an environment variable `CONDA_LIBMAMBA_SOLVER_NO_CHANNELS_FROM_INSTALLED=1`?

Thanks!

**Comment by VitaliyaIoffe** [comment]
Hi @jaimergp, 
I just checked, and yes, I still see 401 auth errors with `conda-libmamba-version 23.5.0` and `CONDA_LIBMAMBA_SOLVER_NO_CHANNELS_FROM_INSTALLED=1`

**Comment by jaimergp** [comment]
Thanks for checking @VitaliyaIoffe!

Let's try a few other things:

- Assuming you are mirroring some external channels in Artifactory, could you please check if [the content written in these pages can help somehow](https://docs.conda.io/projects/conda/en/latest/user-guide/configuration/mirroring.html)?
- Is `conda install --solver=classic` working as expected?
- Conversely, does `mamba` work in your setup?
- Could you please post the whole traceback message (redact as necessary)?

**Comment by VitaliyaIoffe** [comment]
Even conda create env doesn't work with libmamba solver. 
We don't use mamba.
Current setup is working with classic solver.

Whole traceback even edited is not possible - sorry.

```
The following new packages will be installed 
<packages - artifactory paths>

Downloading and extracting packages

CondaHTTPError: HTTP 401 UNAUTHORIZED for url https://artifactory-conda-forge-mirror


```

**Comment by VitaliyaIoffe** [comment]
We don't use mamba.
The current setup is working with a classic solver.

Whole traceback even edited is not possible - sorry.

```
The following new packages will be installed 
<packages - artifactory paths>

Downloading and extracting packages
...
Traceback (most recent call last):
File
"...\lib\site-packages\condalauxlib\logz.py",line 188, in stringify
requests models Response builder (builder, obj)
File
"...\lib\site-packages\conda\auxlib\logz.py'
resp = response _object.json()
File
"..\lib\site-packages\requests\models.py"
line 975, in json
raise RequestsJS0NDecodeError(e.msg, e.doc, e.pos)
requests. exceptions.JS0NDecodeError: Expecting value: line 1 column 1 (char 0)
....
CondaHTTPError: HTTP 401 UNAUTHORIZED for url https://artifactory-conda-forge-mirror/package1.conda
...
CondaHTTPError: HTTP 403 UNAUTHORIZED for url https://artifactory-conda-forge-mirror/package2.tar.bz2
```

I suppose it is important to mention that I use a script, which is rendering a file with all dependencies by `conda render recipe --file filename` and then is updating the current env using `conda env update --file deps`.

**Comment by jaimergp** [comment]
> We don't use mamba.

I am only asking because it would allow us to rule out whether the issue is on the Python layer (which conda-libmamba-solver provides on its own) or in the C++ layer (which we share with mamba). If you see the issue with mamba too, then we need to look into a different part of the code. It'd be useful if you can test it.

Also is there a chance we can see the contents of the script that uses `conda render`?

**Comment by stevendavis** [comment]
I'm also seeing the HTTP 401 UNAUTHORIZED problem with libmamba. We do not use mamba.

Configuration:
```
conda 23.5.0
conda-libmamba-solver 23.5.0
libmamba 1.4.1
```

.condarc file:
```
channel_alias: https://USER:TOKEN@xxx.gov/artifactory/api/conda/all-conda
channels:
  - https://USER:TOKEN@xxx.gov/artifactory/api/conda/all-conda
default_channels:
  - https://USER:TOKEN@xxx.gov/artifactory/api/conda/all-conda
```

Create env command:
`conda env create --file test.yml --force --solver=libmamba`

test.yml file:
```
name: test

dependencies:
  - python=3.9
  - "openssl<3.0" # for compatibility with Artifactory
  - pip
```

Error message:
`CondaHTTPError: HTTP 401 UNAUTHORIZED for url <https://xxx.gov/artifactory/api/conda/all-conda/linux-64/python-3.9.16-h7a1cb2a_2.conda>`

Is there a workaround for this problem?

**Comment by jaimergp** [comment]
I don't have the infra necessary to reproduce the problem so I can only guess some things to try! Right now I can infer that:

- this is happening in the download phase of the artifact (and not the repodata fetching)
- conda-libmamba-solver is to be blamed because `classic` works

Since we are using conda internals for the fetching... it must be how we build the transaction.

Let's try changing this line:

https://github.com/conda/conda-libmamba-solver/blob/d49a438527ef8203c8ba5de18dc5e14b235849e5/conda_libmamba_solver/solver.py#L743

So it looks like this:

```diff
-        kwargs["url"] = join_url(channel_info.noauth_url, pkg_filename)
+        kwargs["url"] = join_url(channel_info.full_url, pkg_filename)
```

Let me know if that works!

**Comment by VitaliyaIoffe** [comment]
I don't see the issue now. But I did a lot of local changes (upgrade openssl to 3.* and migration from API token to Identity token in artifactory). 

Let's wait for @stevendavis answer.

**Comment by stevendavis** [comment]
Good news. I edited the source in my site-packages directory and yes, thank you, the proposed change is working properly with Artifactory.

`kwargs["url"] = join_url(channel_info.full_url, pkg_filename)`

May I ask, would it be possible to release this change soon? My coworkers will also need this change and I don't have an easy way to make the change for other users.

**Comment by jaimergp** [comment]
Yes, we'll have it in the July release. Thank you for trying it out!
